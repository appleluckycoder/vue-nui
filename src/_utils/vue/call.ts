export function call<A extends any[]> (funcs: Function[] | Function, ...args: A): void {
  if (Array.isArray(funcs)) funcs.forEach((func) => call(func, ...args))
  else return funcs(...args)
}

