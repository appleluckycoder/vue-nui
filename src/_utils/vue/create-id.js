export function createId (length = 8) {
  return Math.random().toString(16).slice(2, 2 + length)
}
