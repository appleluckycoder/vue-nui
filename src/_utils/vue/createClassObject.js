export default function createClassObject (classValue) {
  if (typeof classValue === 'string') {
    return classValue
      .split(' ')
      .filter(v => v.length)
      .reduce((prevValue, currentValue) => {
        prevValue[currentValue] = true
        return prevValue
      }, {})
  } else if (classValue && typeof classValue === 'object') {
    return classValue
  }
  return null
}
