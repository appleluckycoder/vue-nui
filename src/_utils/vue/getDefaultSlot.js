export default function (instance, fallback) {
  fallback = fallback === undefined ? [] : fallback
  const defaultScopedSlot = instance.$slots.default
  return (defaultScopedSlot && defaultScopedSlot()) || fallback
}
