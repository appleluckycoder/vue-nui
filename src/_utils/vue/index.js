export * from './src/get-slot'
export * from './src/get-v-node-children'
export * from './src/create-id'
