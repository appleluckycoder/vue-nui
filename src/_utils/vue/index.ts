export { getSlot } from './get-slot'
export { getVNodeChildren } from './get-v-node-children'
export { keep } from './keep'
export { omit } from './omit'
export { flatten } from './flatten'
export { call } from './call'
export { keysOf } from './keysOf'
export { render } from './render'
export type { MaybeArray } from './call'
