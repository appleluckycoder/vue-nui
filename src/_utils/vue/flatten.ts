import { Fragment, VNodeChild } from 'vue'

// o(n) flatten
export function flatten (vNodes: VNodeChild[], result: VNodeChild[] = []) {
  vNodes.forEach((vNode) => {
    if (vNode === null) return
    if (typeof vNode !== 'object') {
      result.push(vNode)
      return
    }
    if (Array.isArray(vNode)) {
      flatten(vNode, result)
      return
    }
    if (vNode.type === Fragment) {
      if (vNode.children === null) return
      if (Array.isArray(vNode.children)) {
        flatten(vNode.children, result)
      }
      // rawSlot
    } else {
      result.push(vNode)
    }
  })
  return result
}
