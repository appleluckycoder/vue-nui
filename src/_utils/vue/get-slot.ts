import { ComponentPublicInstance, VNode } from "vue"

export function getSlot (instance: ComponentPublicInstance, slotName = 'default', fallback: VNode[] = []) {
  const slots = instance.$slots
  const slot = slots[slotName]
  if (slot === undefined) return fallback
  return slot()
}
