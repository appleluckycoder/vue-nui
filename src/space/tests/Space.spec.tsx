import { mount, VueWrapper } from '@vue/test-utils'
import { h } from 'vue'
import { NSpace } from '../index'

const getChildrenNode = (wrapper: VueWrapper<any>): any[] => {
  return (
    wrapper.findAll('div').filter((v) => {
      return !v.classes().includes('n-space')
    }) || []
  )
}

describe('n-space', () => {
  it('should work with import on demand', () => {
    mount(NSpace)
  })

  it('render empty children', () => {
    const wrapper = mount({
      render () {
        return <NSpace />
      }
    })
    expect(wrapper.find('.n-space')).not.toBe(null)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('render space string size', () => {
    const wrapper = mount({
      render () {
        return <NSpace size="large" />
      }
    })
    expect(wrapper.attributes('style')).toContain('margin')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('render space number size', () => {
    const size = 20
    const wrapper = mount({
      render () {
        return <NSpace size={size} />
      }
    })
    expect(wrapper.attributes('style')).toContain(`margin-top: -${size / 2}px`)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('render vertical space', () => {
    const wrapper = mount({
      render () {
        return (
          <NSpace vertical>
            <div>1</div>
            <div>2</div>
          </NSpace>
        )
      }
    })
    expect(wrapper.attributes('style')).toContain('flex-direction: column;')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render with invalidElement', () => {
    const wrapper = mount({
      render () {
        return (
          <NSpace>
            text1<span>text1</span>
            text1
          </NSpace>
        )
      }
    })

    const childNodes = getChildrenNode(wrapper)
    expect(childNodes.length).toBe(3)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('render justify space', () => {
    const justifyList = ['start', 'end'] as Array<'start' | 'end'>
    justifyList.forEach((pos) => {
      const wrapper = mount({
        render () {
          return (
            <NSpace justify={pos}>
              <div>1</div>
              <div>2</div>
            </NSpace>
          )
        }
      })
      expect(wrapper.attributes('style')).toContain(
        `justify-content: flex-${pos};`
      )
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  it('render custom style space', () => {
    const wrapper = mount(
      {
        render () {
          return <NSpace>div</NSpace>
        }
      },
      {
        props: {
          itemStyle: { backgroundColor: 'red', color: 'yellow' }
        }
      }
    )
    const childNodes = getChildrenNode(wrapper)

    expect(childNodes[0].attributes('style')).toContain(
      'background-color: red; color: yellow;'
    )
    expect(wrapper.html()).toMatchSnapshot()
  })
})
