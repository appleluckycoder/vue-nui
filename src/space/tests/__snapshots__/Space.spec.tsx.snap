// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`n-space render custom style space 1`] = `
"<div role=\\"none\\" class=\\"n-space\\" style=\\"display: flex; flex-direction: row; justify-content: flex-start; flex-wrap: wrap; margin-top: -4px; margin-bottom: -4px;\\">
  <div role=\\"none\\" style=\\"background-color: red; color: yellow; max-width: 100%; padding-top: 4px; padding-bottom: 4px;\\">div</div>
</div>"
`;

exports[`n-space render empty children 1`] = `"<div role=\\"none\\" class=\\"n-space\\" style=\\"display: flex; flex-direction: row; justify-content: flex-start; flex-wrap: wrap; margin-top: -4px; margin-bottom: -4px;\\"></div>"`;

exports[`n-space render justify space 1`] = `
"<div role=\\"none\\" class=\\"n-space\\" style=\\"display: flex; flex-direction: row; justify-content: flex-start; flex-wrap: wrap; margin-top: -4px; margin-bottom: -4px;\\">
  <div role=\\"none\\" style=\\"max-width: 100%; margin-right: 12px; padding-top: 4px; padding-bottom: 4px;\\">
    <div>1</div>
  </div>
  <div role=\\"none\\" style=\\"max-width: 100%; padding-top: 4px; padding-bottom: 4px;\\">
    <div>2</div>
  </div>
</div>"
`;

exports[`n-space render justify space 2`] = `
"<div role=\\"none\\" class=\\"n-space\\" style=\\"display: flex; flex-direction: row; justify-content: flex-end; flex-wrap: wrap; margin-top: -4px; margin-bottom: -4px;\\">
  <div role=\\"none\\" style=\\"max-width: 100%; margin-right: 12px; padding-top: 4px; padding-bottom: 4px;\\">
    <div>1</div>
  </div>
  <div role=\\"none\\" style=\\"max-width: 100%; padding-top: 4px; padding-bottom: 4px;\\">
    <div>2</div>
  </div>
</div>"
`;

exports[`n-space render space number size 1`] = `"<div role=\\"none\\" class=\\"n-space\\" style=\\"display: flex; flex-direction: row; justify-content: flex-start; flex-wrap: wrap; margin-top: -10px; margin-bottom: -10px;\\"></div>"`;

exports[`n-space render space string size 1`] = `"<div role=\\"none\\" class=\\"n-space\\" style=\\"display: flex; flex-direction: row; justify-content: flex-start; flex-wrap: wrap; margin-top: -6px; margin-bottom: -6px;\\"></div>"`;

exports[`n-space render vertical space 1`] = `
"<div role=\\"none\\" class=\\"n-space\\" style=\\"display: flex; flex-direction: column; justify-content: flex-start; flex-wrap: nowrap;\\">
  <div role=\\"none\\" style=\\"max-width: 100%; margin-bottom: 8px;\\">
    <div>1</div>
  </div>
  <div role=\\"none\\" style=\\"max-width: 100%;\\">
    <div>2</div>
  </div>
</div>"
`;

exports[`n-space should render with invalidElement 1`] = `
"<div role=\\"none\\" class=\\"n-space\\" style=\\"display: flex; flex-direction: row; justify-content: flex-start; flex-wrap: wrap; margin-top: -4px; margin-bottom: -4px;\\">
  <div role=\\"none\\" style=\\"max-width: 100%; margin-right: 12px; padding-top: 4px; padding-bottom: 4px;\\">text1</div>
  <div role=\\"none\\" style=\\"max-width: 100%; margin-right: 12px; padding-top: 4px; padding-bottom: 4px;\\"><span>text1</span></div>
  <div role=\\"none\\" style=\\"max-width: 100%; padding-top: 4px; padding-bottom: 4px;\\">text1</div>
</div>"
`;
